pymc/step_methods/state.py

```diff
@@ -0,0 +1,98 @@
+#   Copyright 2024 The PyMC Developers
+#
+#   Licensed under the Apache License, Version 2.0 (the "License");
+#   you may not use this file except in compliance with the License.
+#   You may obtain a copy of the License at
+#
+#       http://www.apache.org/licenses/LICENSE-2.0
+#
+#   Unless required by applicable law or agreed to in writing, software
+#   distributed under the License is distributed on an "AS IS" BASIS,
+#   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+#   See the License for the specific language governing permissions and
+#   limitations under the License.
+from copy import deepcopy
+from dataclasses import dataclass, fields
+from typing import Any
+
+import numpy as np
+
+
+class MetaDataClassState(type):
```

Comment:
I already said it somewhere else but I'll add it here too. I wanted to make the step method states simple `dataclass` wrapped classes. I ran into problems revolving around `__eq__` and then also around positional arguments being defined after arguments with default values. To avoid these problems, I had to add some arguments to the `dataclass` decorator and I would sometimes forget them, leading to errors down the line. The simplest solution that didn't involve having to always write down boilerplate code with every `State` definition was to add this metaclass. It simply creates the subclass type and then wraps that using the `dataclass(eq=False, kw_only=True)` decorator. That way, you just need to inherit from `DataClassState` and you'll get the guarantee of working with a `dataclass` that has the specially crafted `__eq__` method.